src/Microsoft.Data/BaseColumn.cs

```diff
@@ -0,0 +1,338 @@
+        /// <param name="invertMapIndices"></param>
+        /// <returns></returns>
+        public virtual BaseColumn Clone(BaseColumn mapIndices = null, bool invertMapIndices = false) => throw new NotImplementedException();
+
+        public virtual BaseColumn Sort(bool ascending = true) => throw new NotImplementedException();
+
+        public virtual BaseColumn GetAscendingSortIndices() => throw new NotImplementedException();
+
+        protected delegate long GetBufferSortIndex(int bufferIndex, int sortIndex);
+        protected delegate T GetValueAtBuffer<T>(int bufferIndex, int valueIndex);
+        protected delegate int GetBufferLengthAtIndex(int bufferIndex);
+        protected void PopulateColumnSortIndicesWithHeap<T>(SortedDictionary<T, List<Tuple<int, int>>> heapOfValueAndListOfTupleOfSortAndBufferIndex, PrimitiveColumn<long> columnSortIndices, GetBufferSortIndex getBufferSortIndex, GetValueAtBuffer<T> getValueAtBuffer, GetBufferLengthAtIndex getBufferLengthAtIndex)
+        {
+            while (heapOfValueAndListOfTupleOfSortAndBufferIndex.Count > 0)
+            {
+                KeyValuePair<T, List<Tuple<int, int>>> minElement = heapOfValueAndListOfTupleOfSortAndBufferIndex.ElementAt(0);
+                T value = minElement.Key;
+                List<Tuple<int, int>> tuplesOfSortAndBufferIndex = minElement.Value;
+                Tuple<int, int> sortAndBufferIndex;
+                if (tuplesOfSortAndBufferIndex.Count == 1)
+                {
+                    heapOfValueAndListOfTupleOfSortAndBufferIndex.Remove(minElement.Key);
+                    sortAndBufferIndex = tuplesOfSortAndBufferIndex[0];
+                }
+                else
+                {
+                    sortAndBufferIndex = tuplesOfSortAndBufferIndex[tuplesOfSortAndBufferIndex.Count - 1];
+                    tuplesOfSortAndBufferIndex.RemoveAt(tuplesOfSortAndBufferIndex.Count - 1);
+                }
+                int sortIndex = sortAndBufferIndex.Item1;
+                int bufferIndex = sortAndBufferIndex.Item2;
+                long bufferSortIndex = getBufferSortIndex(bufferIndex, sortIndex);
+                columnSortIndices.Append(bufferSortIndex);
+                if (sortIndex + 1 < getBufferLengthAtIndex(bufferIndex))
+                {
+                    int nextSortIndex = sortIndex + 1;
+                    T nextValue = getValueAtBuffer(bufferIndex, nextSortIndex);
+                    heapOfValueAndListOfTupleOfSortAndBufferIndex.Add(nextValue, new List<Tuple<int, int>>() { new Tuple<int, int>((int)nextSortIndex, bufferIndex) });
+                }
+            }
+
+        }
+        private int FloorLog2PlusOne(int n)
+        {
+            Debug.Assert(n >= 2);
+            int result = 2;
+            n >>= 2;
+            while (n > 0)
+            {
+                ++result;
+                n >>= 1;
+            }
+            return result;
+        }
+
+        protected void IntrospectiveSort<T>(
```

Comment:
`internal`. More implementation detail we don't want in the external API.

Reply:
Why isn't this static?

---
In reply to: [283058030](https://github.com/dotnet/corefxlab/pull/2660#discussion_r283058030) [](ancestors = 283058030)